File: rc4_java/Rc4.java
package rc4_java;

public class Rc4 {
	private static int mod(int i, int j)
	{
		int r;
		if(i>j)
			r=i%j;
		else if(i<j)
			r=i;
		else
			r=0;
		return r;
	}
	public static void  rc4_init(short s[],String keys,int len) {
		int i = 0;
		int j = 0;
		short[] k= new short[256];
		short tmp = 0;
		byte[] key=keys.getBytes();
		for ( i=0;i<256;i++)
		{
			s[i]= (short) i;
			k[i]=key[mod(i,len)];
		}
		for( i=0;i<256;i++)
		{
			j= mod((int)(j+s[i]+k[i]),256);
			tmp=s[i];
			s[i] = s[j];
			s[j]=tmp;
		}
	}
	public static void  rc4_crypt(short s[],short data[],int len) {
		int i=0;
		int j=0;
		int t=0;
		int k=0;
		short tmp;
		for(k=0;k<len;k++)
		{
			i=mod((i+1),256);
			j=mod((j+s[i]),256);
			tmp=s[i];
			s[i]=s[j];
			s[j]=tmp;
			t=mod((s[i]+s[j]),256);
			data[k]^=s[t];
		}
	}
}

File: rc4_c/rc4.c
#include<stdio.h>
#include<string.h>

void rc4_init(unsigned char *s, unsigned char *key, unsigned long Len)
{
    int i =0;
    int j =0;
    char k[256] = {0};
    unsigned char tmp = 0;
    for(i=0;i<256;i++)
    {
        s[i]=i;
        k[i]=key[i%Len];
    }
    for(i=0; i<256; i++)
    {
        j=(j+s[i]+k[i])%256;
        tmp = s[i];
        s[i] = s[j];
        s[j] = tmp;
    }
}

void rc4_crypt(unsigned char *s, unsigned char *Data, unsigned long Len)
{
    int i = 0;
    int j = 0;
    int t = 0;
    unsigned long k = 0;
    unsigned char tmp;
    for(k=0;k<Len;k++)
    {
        i=(i+1)%256;
        j=(j+s[i])%256;
        tmp = s[i];
        s[i] = s[j];
        s[j] = tmp;
        t=(s[i]+s[j])%256;
        Data[k] ^= s[t];
    }
}

void main()
{
    unsigned char s1[256] = {0};
    unsigned char s2[256] = {0};
    char key[256] = {"test"};
    char pData[6] = {0x6f, 0xc1, 0xbe, 0xe2, 0x2e, 0x40};
    int i;
    unsigned long len = 6;
    printf("key = %s, length = %d\n",key,strlen(key));
    printf("data:");
    for(i=0;i<len;i++)
        printf("0x%2x,",*(pData+i));
    printf("\n");
    //encode
    rc4_init(s1,(unsigned char *)key,strlen(key));
    rc4_crypt(s1,(unsigned char *)pData,len);
    printf("encry:");
    for(i=0;i<len;i++)
        printf("0x%2x,",*(pData+i));
    printf("\n");
    //decode
    rc4_init(s2,(unsigned char *)key, strlen(key));
    rc4_crypt(s2,(unsigned char *)pData,len);
    printf("decry:");
    for(i=0;i<len;i++)
        printf("0x%2x,",*(pData+i));
    printf("\n");
}
